README.md
<p align="center"><img src="https://user-images.githubusercontent.com/72305598/184533504-d38cfa2d-97b0-4dd1-95a2-d4a2a2dfb576.png" alt="Jeve" /></p>

<p align="center">
<img src="https://img.shields.io/endpoint.svg?url=https%3A%2F%2Factions-badge.atrox.dev%2Fjack-carling%2Fjeve%2Fbadge%3Fref%3Dmain&style=flat" />
<img src="https://img.shields.io/npm/dt/jeve" />
<img src="https://img.shields.io/npm/v/jeve" />
<img src="https://img.shields.io/npm/l/jeve" />
</p>
<br />

Jeve is a JavaScript framework for effortlessly building an API with a [self-written documentation](#self-written-documentation). Powered by [Express](https://expressjs.com/) and [Mongoose](https://mongoosejs.com/) for native [MongoDB](https://www.mongodb.com/) support. Inspired by [PyEve](https://python-eve.org/) for Python.

## Table of contents

-   [Install](#install)
-   [Quick start](#quick-start)
-   [Settings](#settings)
-   [Schema](#schema)
-   [Resource methods](#resource-methods)
-   [Item methods](#item-methods)
-   [Validations](#validations)
-   [Params & queries](#params--queries)
-   [Middleware](#middleware)
-   [Custom routes](#custom-routes)
-   [Accessing models](#accessing-models)
-   [Self-written documentation](#self-written-documentation)

## Install

```
$ npm install jeve
```

## Quick start

Jeve runs on port 5000 and uses `mongodb://localhost` as the default MongoDB connection string URI. Presuming that the database is running locally on the default port the following code is the bare minimum to get the API up and running:

```javascript
const Jeve = require('jeve')

const settings = {
    domain: {
        people: {},
    },
}

const jeve = new Jeve(settings)
jeve.run()
```

That's all it takes for the API to go live with an endpoint. We can now try to **GET** `/people`.

```
$ curl -i http://localhost:5000/people
HTTP/1.1 204 No Content
```

We're live but the endpoint returns no content since we haven't set up a schema. More about that later.

## Settings

In order to change the default port and MongoDB URI simply add the keys to the root of the `settings` object. They can also be added in a `.env` file. In case values exists both in `.env` and in the root of the `settings` object, the root value will be superseded by dotenv.

| key      | value  | default               | dotenv   |
| -------- | ------ | --------------------- | -------- |
| port     | number | `5000`                | PORT     |
| database | string | `mongodb://localhost` | DATABASE |

If we would want to run Jeve on port 5100 instead for example:

```javascript
const settings = {
    domain: {
        people: {},
    },
    port: 5100,
}
```

## Schema

Jeve will automatically create Mongoose Models based on a `schema` object under each domain. Let's add name and age to the people domain as types string and number.

```javascript
const settings = {
    domain: {
        people: {
            schema: {
                name: 'string',
                age: 'number',
            },
        },
    },
}
```

Types can be written directly as a string:

```javascript
name: 'string'
```

Or as an object with the `type` key:

```javascript
name: {
  type: 'string',
}
```

The latter is needed if other validations like `required` or `unique` are going to be added. In our example, name is a required field.

```javascript
name: {
  type: 'string',
  required: true,
}
```

The following types are supported:

-   string
-   number
-   date
-   boolean
-   objectid
-   object
-   array

We now have a `settings` object which looks like this:

```javascript
const settings = {
    domain: {
        people: {
            schema: {
                name: {
                    type: 'string',
                    required: true,
                },
                age: 'number',
            },
        },
    },
}
```

## Resource methods

Our endpoint `/people` allows HTTP method **GET** by default. In order to save a person to the database, we need to add the **POST** method to our domain resource.

Resource methods are added as strings in an array on the same root as our `schema` object. Valid HTTP methods are **GET** and **POST**.

Before adding this to our settings, let's try to **POST**.

```
$ curl -i -X POST http://localhost:5000/people
HTTP/1.1 404 Not Found
```

As expected, we get a 404 status code response. Add the following to the `people` object and try again:

```javascript
people: {
  resourceMethods: ['GET', 'POST'],
  schema: { ... },
}
```

```
$ curl -i -X POST http://localhost:5000/people
HTTP/1.1 400 Bad Request
```

This time we get a 400 status code response instead along with a json error message since we sent an empty body request.

```javascript
{
  "_success": false,
  "_issues": [
    {
      "name": "required field"
    }
  ]
}
```

If we instead send a proper **POST** application/json with a name:

```
curl -i -d '{"name":"James Smith"}' -H "Content-Type: application/json" -X POST http://localhost:5000/people
HTTP/1.1 201 Created
```

Now our first successful **POST** was made!

```javascript
{
  "_success": true,
  "_item": {
    "_id": "62eebccf0c5aa6efc2d8ceed",
    "name": "James Smith",
    "_created": "2022-08-06T19:11:11.627Z",
    "_updated": "2022-08-06T19:11:11.627Z"
  }
}
```

By default only **GET** methods are allowed unless an array of `resourceMethods` have been defined. If however, you'd like an endpoint only serving **POST** requests, simply add that as the single value to the array.

## Item methods

In the previous example our request returned an item with an `_id`. If we wanted to access only this item in a **GET** request, we could add the `_id` as a parameter to the request: `/people/62eebccf0c5aa6efc2d8ceed`.

By default the only valid HTTP method is **GET**, however if we would want other methods to be allowed we simply add them to our `itemMethods` array in a similar way as the `resourceMethods`.

The main difference to think about is that resource methods take care of the domain itself, accessing `/people` in order to **GET** a list of documents or **POST** a new document. While item methods handle a mandatory parameter which is the `_id` of the document in order to **GET** that specific document or handle updates or deletions. Valid methods are:

-   GET
-   PUT
-   PATCH
-   DELETE

## Validations

Schema keys are actual field names and in case the value is an object instead of the `type` as a string the following validation rules are can be used:

<table>
  <tr>
    <td><code>type</code></td>
    <td>
      Field data type. Required in the schema and can be one of the following:
      <ul>
        <li><code>string</code></li>
        <li><code>number</code></li>
        <li><code>date</code></li>
        <li><code>boolean</code></li>
        <li><code>objectid</code></li>
        <li><code>object</code></li>
        <li><code>array</code></li>
      </ul>
    </td>
  </tr>
  <tr>
    <td><code>required</code></td>
    <td>If <code>true</code> the field is mandatory on insertion.</td>
  </tr>
  <tr>
    <td><code>unique</code></td>
    <td>If <code>true</code> the value of the field must be unique within the collection.</td>
  </tr>
  <tr>
    <td><code>default</code></td>
    <td>
      The default value for the field. When serving POST and PUT requests, missing fields will be assigned the
      configured default values.
    </td>
  </tr>
  <tr>
    <td><code>minLength</code>, <code>maxLength</code></td>
    <td>Minimum and maximum length allowed for <code>string</code> types.</td>
  </tr>
  <tr>
    <td><code>min</code>, <code>max</code></td>
    <td>Minimum and maximum values allowed for <code>number</code> types.</td>
  </tr>
</table>

An example where name has to contain at least 2 characters and email must be unique within the collection:

```javascript
{ /* ... */
  name: {
    type: 'string',
    required: true,
    minLength: 2
  },
  email: {
    type: 'string',
    required: true,
    unique: true,
  }
}
```

## Params & queries

If we make a new **GET** request to `/people`, we now get everything that's stored in the database.

```
$ curl -i http://localhost:5000/people
HTTP/1.1 200 OK
```

```javascript
{
  "_success": true,
  "_items": [
    {
      "_id": "62eebccf0c5aa6efc2d8ceed",
      "name": "James Smith",
      "_created": "2022-08-06T19:11:11.627Z",
      "_updated": "2022-08-06T19:11:11.627Z"
    }
  ],
  "_meta": {
    "total": 1,
    "limit": 10,
    "page": 1,
    "pages": 1
  }
}
```

In our case the `_items` array only contains one (1) object, the one we just added. Responses are paginated by default and the `_meta` object contains information about the specific endpoint.

| \_meta |  description                              |
| ------ | ----------------------------------------- |
| total  | The total number of documents found       |
| limit  | Max results per page                      |
| page   | The page which the cursor is currently on |
| pages  | The total number of pages                 |

If we imagined that we had 12 documents in the `/people` collection the `_meta` response would look something like this:

```javascript
{
  "_success": true,
  "_items": [ /* ... */ ],
  "_meta": {
    "total": 12,
    "limit": 10,
    "page": 1,
    "pages": 2
  }
}
```

Since we know there's a second page, we can simply do a new **GET** with the page query:

```
$ curl -i "http://localhost:5000/people?page=2"
HTTP/1.1 200 OK
```

If we wanted more results per page:

```
$ curl -i "http://localhost:5000/people?limit=20"
HTTP/1.1 200 OK
```

If we're looking for a specific document the `_id` of that document needs to follow as a parameter, for example `/people/62eebccf0c5aa6efc2d8ceed`. This is the way **PATCH**, **PUT** and **DELETE** knows what document to handle as well.

Other valid queries are `sort`, `where` and `select`.

If we wanted our result to be sorted by their creation date we could send the `/people?sort=_created` query as an example. Or if we wanted to reverse the search, simply add `-` before the key value: `/people?sort=-_created`.

The last two parameters accepts json-input, `where` will filter the request. If for example we only wanted a list of people older than 18 we could use the following query: `/people?where={"age":{"$gte": 18}}`. `select` will filter the documents, as in including specific fields or excluding others. If we for example weren't interested in the ages, we could exclude that field by specifying the key along with a `0`: `/people?select={"age":0}`.

## Middleware

Each domain accepts a `preHandler` function which will run before the request. Use cases range from authorization to catching data and manipulating the body. As an example, let's imagine we had a `boolean` value for the field `isAdult` in our schema. We're not sending this value in our request, but we want our middleware to catch it.

```javascript
{ /* ... */
  people: {
    resourceMethods: ['GET', 'POST'],
    schema: {
      age: 'number',
      isAdult: 'boolean',
    },
    preHandler: checkIfAdult,
  }
}
```

In our middleware function `checkIfAdult`, we would simply add the value to it. Don't forget to call with `next()`...

```javascript
function checkIfAdult(req, res, next) {
    const age = req.body?.age
    if (age) req.body.isAdult = age >= 18
    next()
}
```

## Custom routes

Jeve supports custom routes:

| method | function        |
| ------ | --------------- |
| GET    | `jeve.get()`    |
| POST   | `jeve.post()`   |
| PUT    | `jeve.put()`    |
| PATCH  | `jeve.patch()`  |
| DELETE | `jeve.delete()` |

A simple example of a `/greeting` route that returns the text `Hello World!`:

```javascript
jeve.get('/greeting', (req, res) => {
    res.send('Hello World!')
})
```

If `greeting` exists in our `domain` object, the custom route will be skipped and not initialized due to conflict and a message will be shown in the console. However, if the path is deeper, for example `/greeting/swedish`, the custom route will be created.

## Accessing Models

Every model that's dynamically created by Jeve is accessible from the `jeve.model()` function. If we for example wanted to access a model from a custom route and use any native Mongoose function with it:

```javascript
jeve.get('/greeting/:id', async (req, res) => {
    const { id } = req.params
    const person = await jeve.model('people').findOne({ _id: id })
    res.send(`Hello ${person.name}`)
})
```

## Self-written documentation

Jeve will dynamically create it's own documentation and the UI is accessible at `/docs` in the browser. The documentation contains all available routes in the `settings` object and will show which resource and item methods they're accessible by. The accordion contains the `schema` object, an overview of keys and validations.

![](https://user-images.githubusercontent.com/72305598/188952457-6b0a5e6b-03e5-4748-a372-a134293ea833.png)


CHANGELOG.md
## 0.2.2 (2022-09-14)

### Bug fixes

`TypeError: app[route] is not a function`

## 0.2.1 (2022-09-07)

### Features

-   Added handling of `itemMethods` in UI
-   Basic overview of schema and validations in UI

## 0.2.0 (2022-09-04)

### Breaking changes

-   HTTP methods are no longer only defined in the `resourceMethods` array. The values in `resourceMethods` defines which methods are supported at the resource endpoint, supported values are **GET** and **POST**. The `itemMethods` array contains which methods are allowed at the item endpoint: `/resource/id` which is the `_id` of the document. Supported `itemMethods` values are:
    -   GET
    -   PUT
    -   PATCH
    -   DELETE

### Features

-   Added `itemMethods`
-   HTTP method **GET** is no longer added to `resourceMethods` or `itemMethods` by default
    in order to enabled the creation of restricted endpoints

## 0.1.6 (2022-09-03)

### Features

-   Added `object` as valid type in schema

## 0.1.5 (2022-08-28)

### Features

-   Added `min`, `max`, `minLength` and `maxLength` as valid data validation keys

## 0.1.4 (2022-08-27)

### Features

-   Improved casting and validations

## 0.1.3 (2022-08-26)

### Features

-   Validating keys in schema
-   Adding `default` as valid data validation key

## 0.1.2 (2022-08-24)

### Bug fixes

-   Handling objects inside an array in a schema

## 0.1.1 (2022-08-23)

### Bug fixes

-   Remove certain private class functions

## 0.1.0 (2022-08-23)

### Bug fixes

-   Fixed how assets are handled by express static

## 0.0.9 (2022-08-17)

### Bug fixes

-   Include files when publishing to npm

## 0.0.8 (2022-08-17)

### Features

-   Added files for documentation

## 0.0.7 (2022-08-17)

### Code refactoring

-   Fixed tests

## 0.0.6 (2022-08-16)

### Features

-   Export schema as JSON in preparation for documentation

## 0.0.5 (2022-08-16)

### Features

-   Custom routes

## 0.0.4 (2022-08-14)

### Features

-   Middleware in schema with the `preHandler` hook

### Other

-   Added license

## 0.0.3 (2022-08-14)

### Code refactoring

-   New tests
-   Added private class functions

## 0.0.2 (2022-08-13)

### Code refactoring

-   Promises to ensure initialization
-   Started integration tests

## 0.0.1 (2022-08-12)

### Features

-   Chalk logging
-   Dynamically created routes
-   Dynamically created Mongoose Schemas
-   Validations on initialization
-   Handle GET, POST, PUT, PATCH & DELETE


lib/index.js
require('dotenv').config()
const express = require('express')
const chalk = require('chalk')
const mongoose = require('mongoose')
const fs = require('fs')
const path = require('path')
const app = express()

app.use(express.json())
app.use('/docs', express.static(path.join(__dirname, '..', '/jeve')))
app.use('/assets', express.static(path.join(__dirname, '..', '/jeve/assets')))
app.use('/docs', express.static('docs'))
app.set('etag', false)

const { handleGet, handlePost, handlePut, handlePatch, handleDelete } = require('./route')
const { createModelForName } = require('./models')
const { validType, isEmptyObject } = require('./helpers')

class Jeve {
    constructor(settings) {
        this.settings = settings
        this.models = {}
    }

    // Logs
    log(...text) {
        console.log(chalk.gray(`[jeve] ${text}`))
    }

    warning(text) {
        console.log(chalk.yellow(`[jeve] warning - ${text}`))
    }

    error(text) {
        console.log(chalk.red(`[jeve] exit - ${text}`))
    }

    success(text) {
        console.log(chalk.green(`[jeve] ${text}`))
    }

    // Validation
    #validTypeLog(variable, name, type, error = false) {
        const valid = validType(variable, type)
        if (!valid && error) {
            this.error(`${name} needs to be "${type}"`)
        } else if (!valid) {
            this.warning(`${name} needs to be "${type}"`)
        }
        return valid
    }

    #validateCustomRoute(args, functionName) {
        let message = `jeve.${functionName}`
        if (!args.length) {
            message += '()'
            this.warning(`${message} is missing arguments`)
            return false
        } else {
            if (validType(args[0], 'string')) {
                message += `('${args[0]}'`
            } else {
                message += `(${args[0]}`
            }
            if (args.length > 1) {
                message += `, ...)`
            } else {
                message += `)`
            }
        }

        if (!this.#validTypeLog(args[0], `${message} has invalid first argument,`, 'string')) {
            return false
        }

        if (!args[0].includes('/')) {
            this.warning(`${message} has invalid first argument, string needs to contain "/"`)
            return false
        }
        const names = args[0].split('/').filter((v) => v.length)
        const domain = this.settings?.domain

        if (names.length === 1 && domain) {
            if (names[0] in domain) {
                this.warning(
                    `conflict with custom route: \`/${names[0]}\` already exists in settings and will be skipped`
                )
                return false
            }
        }

        return true
    }

    // Run
    async run() {
        console.time('[jeve] initialization')
        this.success(`${new Date().toLocaleTimeString()} - starting jeve...`)
        if (!this.settings) return this.error('missing settings')
        this.#JSON_()
        const port = (process.env.PORT || this.settings.port) ?? 5000
        if (!this.#validTypeLog(port, 'port', 'number', true)) return
        const database = (process.env.DATABASE || this.settings.database) ?? 'mongodb://localhost'
        if (!this.#validTypeLog(database, 'database', 'string', true)) return
        try {
            await mongoose.connect(database)
            this.success('connected to database')
        } catch (e) {
            return this.error(e)
        }
        this.success('initializing...')
        await this.#initialize()
        app.listen(port, () => {
            this.success(`running on port ${port}`)
        })
        console.timeEnd('[jeve] initialization')
    }

    // Create JSON API docs
    #JSON_() {
        const data = JSON.stringify(this.settings)
        if (!fs.existsSync('docs')) {
            fs.mkdirSync('docs')
        }
        if (!fs.existsSync('docs/api.json')) {
            this.#writeJSON(data)
        } else {
            const fileData = fs.readFileSync('docs/api.json', { encoding: 'utf-8' })
            if (fileData !== data) {
                this.#writeJSON(data)
            }
        }
    }

    #writeJSON(data) {
        fs.writeFileSync('docs/api.json', data, 'utf-8')
    }

    // Initialization
    #initialize() {
        const domain = this.settings.domain
        return new Promise(async (resolve) => {
            if (!this.#validTypeLog(domain, 'domain', 'object')) return
            for (const d in domain) {
                if (!this.#validTypeLog(domain[d], `domain.${d}`, 'object')) continue

                const resourceMethods = this.settings.domain[d].resourceMethods ?? ['GET']
                const itemMethods = this.settings.domain[d].itemMethods ?? ['GET']
                if (!this.#validTypeLog(resourceMethods, `domain.${d}.resourceMethods`, 'array')) continue
                if (!this.#validTypeLog(itemMethods, `domain.${d}.resourceMethods`, 'array')) continue

                const resourceMethodsMap = resourceMethods.map((v, i) => {
                    if (!this.#validTypeLog(v, `domain.${d}.resourceMethods.${i}`, 'string')) {
                        return v.toUpperCase()
                    } else {
                        return v
                    }
                })
                const itemMethodsMap = itemMethods.map((v, i) => {
                    if (!this.#validTypeLog(v, `domain.${d}.itemMethods.${i}`, 'string')) {
                        return v.toUpperCase()
                    } else {
                        return v
                    }
                })
                const validMethods = []
                for (const r of resourceMethodsMap) {
                    const validResourceMethods = ['GET', 'POST']
                    if (!validResourceMethods.includes(r)) {
                        this.warning(`domain.${d}.resourceMethods contains unknown value "${r}"`)
                        continue
                    }
                    validMethods.push(r)
                }
                for (const i of itemMethodsMap) {
                    const validItemMethods = ['GET', 'PUT', 'PATCH', 'DELETE']
                    if (!validItemMethods.includes(i)) {
                        this.warning(`domain.${d}.itemMethods contains unknown value "${i}"`)
                        continue
                    }
                    validMethods.push(i)
                }
                app.use(`/${d}/:id?`, (req, res, next) => {
                    const { id } = req.params
                    if (id && !validType(id, 'objectid')) {
                        return res.sendStatus(404)
                    }
                    if (id && !itemMethodsMap.includes(req.method)) {
                        return res.sendStatus(404)
                    }
                    if (!id && !resourceMethodsMap.includes(req.method)) {
                        return res.sendStatus(404)
                    }
                    next()
                })
                const combinedMethods = Array.from(new Set(validMethods))
                for (const m of combinedMethods) {
                    let preHandler = domain[d].preHandler
                    if (preHandler) {
                        if (!this.#validTypeLog(preHandler, `domain.${d}.preHandler`, 'function')) {
                            preHandler = (req, res, next) => {
                                next()
                            }
                        }
                    } else {
                        preHandler = (req, res, next) => {
                            next()
                        }
                    }
                    await this.#createRoute(m.toLowerCase(), d, preHandler)
                }
                this.log(`created route(s) for - ${d}`)
                const schema = this.settings.domain[d].schema
                const validSchema = this.#validTypeLog(schema, `domain.${d}.schema`, 'object')
                if (!validSchema) continue
                if (isEmptyObject(schema)) {
                    this.warning(`domain.${d}.schema is empty`)
                    continue
                }
                await this.createModelForName(d)
            }
            resolve()
        })
    }

    // Create Routes
    #createRoute(route, domain, preHandler) {
        return new Promise((resolve) => {
            app[route](`/${domain}/:id?`, preHandler, (req, res) => {
                if (req.method === 'GET') {
                    this.handleGet(req, res, domain)
                } else if (req.method === 'POST') {
                    this.handlePost(req, res, domain)
                } else if (req.method === 'PUT') {
                    this.handlePut(req, res, domain)
                } else if (req.method === 'PATCH') {
                    this.handlePatch(req, res, domain)
                } else if (req.method === 'DELETE') {
                    this.handleDelete(req, res, domain)
                } else {
                    res.status(501).json({ _success: false, _error: `Method \`${req.method} is not yet implemented.` })
                }
            })
            resolve()
        })
    }

    // Custom routes
    get(...args) {
        if (this.#validateCustomRoute(args, 'get')) return app.get(...args)
    }

    post(...args) {
        if (this.#validateCustomRoute(args, 'post')) return app.post(...args)
    }

    put(...args) {
        if (this.#validateCustomRoute(args, 'put')) return app.put(...args)
    }

    patch(...args) {
        if (this.#validateCustomRoute(args, 'patch')) return app.patch(...args)
    }

    delete(...args) {
        if (this.#validateCustomRoute(args, 'delete')) return app.delete(...args)
    }

    // Access models
    model(model) {
        if (model && this.models[model]) {
            return this.models[model]
        }
        return undefined
    }
}

Jeve.prototype.handleGet = handleGet
Jeve.prototype.handlePost = handlePost
Jeve.prototype.handlePut = handlePut
Jeve.prototype.handlePatch = handlePatch
Jeve.prototype.handleDelete = handleDelete
Jeve.prototype.createModelForName = createModelForName

module.exports = Jeve
